src/data/qpf/multivariate/constructions/cofix.lean

```diff
@@ -0,0 +1,433 @@
+/-
+Copyright (c) 2018 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jeremy Avigad, Simon Hudon
+-/
+
+import control.functor.multivariate
+import data.pfunctor.multivariate.basic
+import data.pfunctor.multivariate.M
+import data.qpf.multivariate.basic
+
+/-!
+# The final co-algebra of a multivariate qpf is again a qpf.
+
+Similarly to the initial algebra of a multivariate qpf,
+we define the corresponding polynomial functor by selecting:
+ * A := a tree-like structure without information in the nodes
+ * B := given the tree-like structure `t`, `B t` is a valid path
+   from the root of `t` to any given node.
+
+As a result `P.obj α` is made of a dataless tree and a function from
+its valid paths to values of `α`
+
+The difference with the polynomial functor of an initial algebra is
+that `A` is a possibly infinite tree
+
+-/
+
+universe u
+open_locale mvfunctor
+
+namespace mvqpf
+open typevec mvpfunctor
+open mvfunctor (liftp liftr)
+
+variables {n : ℕ} {F : typevec.{u} (n+1) → Type u} [mvfunctor F] [q : mvqpf F]
+include q
+
+/-- Corecursion on M-types -/
+def corecF {α : typevec n} {β : Type*} (g : β → F (α.append1 β)) : β → q.P.M α :=
+M.corec _ (λ x, repr (g x))
+
+theorem corecF_eq {α : typevec n} {β : Type*} (g : β → F (α.append1 β)) (x : β) :
+  M.dest q.P (corecF g x) = append_fun id (corecF g) <$$> repr (g x) :=
+by rw [corecF, M.dest_corec]
+
+/-- Characterization of desirable equivalence relations on M-types -/
+def is_precongr {α : typevec n} (r : q.P.M α → q.P.M α → Prop) : Prop :=
+  ∀ ⦃x y⦄, r x y →
+    abs (append_fun id (quot.mk r) <$$> M.dest q.P x) =
+      abs (append_fun id (quot.mk r) <$$> M.dest q.P y)
+
+/-- Equivalence relation on M-types representing a value of type `cofix F` -/
+def Mcongr {α : typevec n} (x y : q.P.M α) : Prop :=
+∃ r, is_precongr r ∧ r x y
+
+/-- Greatest fixed point of functor F. The result is a functor with one fewer parameters
+than the input. For `F a b c` a ternary functor, fix F is a binary functor such that
+
+```lean
+cofix F a b = F a b (cofix F a b)
+```
+-/
+def cofix (F : typevec (n + 1) → Type u) [mvfunctor F] [q : mvqpf F] (α : typevec n) :=
+quot (@Mcongr _ F _ q α)
+
+instance {α : typevec n} [inhabited q.P.A] [Π (i : fin2 n), inhabited (α i)] :
+  inhabited (cofix F α) := ⟨ quot.mk _ (default _) ⟩
+
+/-- maps every element of the W type to a canonical representative -/
+def Mrepr {α : typevec n} : q.P.M α → q.P.M α := corecF (abs ∘ M.dest q.P)
+
+/-- `cofix F` is itself a functor -/
```

Comment:
This should be a description, not a statement: "the `map` function for the functor `cofix F`."